File: pesquisa-em-largura/pesquisa-em-largura.js
/*
* | IMPLEMENTAÇÃO DO ALGORITMO |
? 1. Crie uma fila contendo todas as pessoas que devem ser verificadas
? 2. Retire uma pessoa da fila
? 3. Confira se esta pessoa é um vendendor de mangas
? 4. É um vendedor de manga? Pronto, finaliza o algoritmo. Não é? Adicione todos os vizinhos dessa pessoa na lista
? 5. Repita
? 6. Se a lista estiver vazio signiifica que não existem vendedores de manga nessa rede
*/

const grafo = {}

grafo['voce'] = ['alice', 'bob', 'claire']
grafo['bob'] = ['anuj', 'peggy']
grafo['alice'] = ['peggy']
grafo['claire'] = ['thom', 'jonny']
grafo['anuj'] = []
grafo['peggy'] = []
grafo['thom'] = []
grafo['jonny'] = []

const eVendedor = (nome) => {
    return nome[nome.length - 1] === 'm'
}

function pesquisaEmLargura(nome){
    let filaDePesquisa = []
    filaDePesquisa = filaDePesquisa.concat(grafo[nome])

    const pesquisados = []
    while (filaDePesquisa.length){
      let pessoa = filaDePesquisa.shift()
      
      if (pesquisados.indexOf(pessoa) === -1){
        if (eVendedor(pessoa)){
          console.log(`${pessoa} é um(a) vendedor(a) de manga!`)
          return true
        }
  
        filaDePesquisa = filaDePesquisa.concat(grafo[pessoa])
        
        pesquisados.push(pessoa)
      }
    }
    return false
}
  
pesquisaEmLargura('voce')

File: pesquisa-em-largura/pesquisa-em-largura.py
grafo = {}

grafo['voce'] = ['alice', 'bob', 'claire']
grafo['bob'] = ['anuj', 'peggy']
grafo['alice'] = ['peggy']
grafo['claire'] = ['thom', 'jonny']
grafo['anuj'] = []
grafo['peggy'] = []
grafo['thom'] = []
grafo['jonny'] = []

from collections import deque #? deque - double ended queue (fila com dois finais)
fila_de_pesquisa = deque()

fila_de_pesquisa += grafo['voce']

def pessoa_e_vendedor(nome): #* Verifica se o nome da pessoa termina com a letra m
    return nome[-1] == 'm'

def pesquisa(nome):
    fila_de_pesquisa = deque()
    fila_de_pesquisa += grafo[nome]
    verificados = []
    while fila_de_pesquisa: #? Enquanto a fila não estiver vazia
        pessoa = fila_de_pesquisa.popleft() #? Pega a primeira pessoa da fila
        if not pessoa in verificados:
            if pessoa_e_vendedor(pessoa): #? Verifica se a pessoa é uma vendedora de mangas
                print(pessoa + ' é um vendedor de manga!')
            else:
                fila_de_pesquisa += grafo[pessoa]
                verificados.append(pessoa)

pesquisa('voce')